assets/mdk/proof-dependencies.mdk

```diff
@@ -0,0 +1,87 @@
+
+<!-- inline-dafny unspecified/BinarySearch -->
+
+Dafny includes a verification logger that will describe the status of and statistics about each verification goal in the program. When proof dependency analysis is enabled, this will include information about which potential proof dependencies did or did not take place in the actual proof. This information will only be included if one of the other flags enabling proof dependency analysis is enabled, however. So, if we analyze the binary search example using the command `dafny verify --log-format text --warn-redundant-assertions`, the output will include the following text.
+
+```
+    Unused by proof:
+      unspecified.dfy(4,14)-(4,28): assignment (or return)
+      unspecified.dfy(4,14)-(4,28): assignment (or return)
+      unspecified.dfy(6,15)-(6,34): loop invariant
+      unspecified.dfy(10,7)-(10,16): assignment (or return)
+      unspecified.dfy(14,7)-(14,17): assignment (or return)
+      unspecified.dfy(17,3)-(17,12): assignment (or return)
+      unspecified.dfy(18,1)-(18,1): out-parameter 'r', which is subject to definite-assignment rules, is always initialized at this return point
+```
+
+This shows that the assignment/return statements on lines 10, 14, and 17, for example, were not necessary to show that the method is well-defined.
+
+
+Enabling proof dependency calculation requires additional work as part of the proof process. This means it can slow down verification, and potentially cause especially brittle proofs to fail. In our experiments so far, it can add ~30% to verification time. If your project is developed with the feature turned on from the start, however, this cost can be mitigated, especially when taking [verification optimization](https://dafny.org/latest/VerificationOptimization/VerificationOptimization) guidance into account.
+
+# Theory and implementation
+
+Internally, Dafny's proof dependency analysis is built on the common SMT feature of _unsatisfiable cores_. Dafny encodes each verification goal as an SMT query that negates the original goal. This means that a conclusion that the negated goal is unsatisfiable (_i.e._, no value exists that will make it true) means that the original goal is valid (_i.e._, true for all possible values). Many SMT solvers, including the Z3 solver that Dafny uses by default, can accompany a conclusion of "unsatisfiable" with a subset of the sub-expressions (clauses, in SMT terminology) from the original goal that is still unsatisfiable. This subset is generally smaller than the original formula, though it is not guaranteed to be so, and is not guaranteed to be minimal.
+
+Because unsatisfiable cores are not guaranteed to be minimal, Dafny may sometimes fail to warn about some goals that are proved using a contradiction, or some assumptions that are not ultimately necessary. In our early experience, however, Dafny does produce useful warnings for all large code bases we've tried it on.
```

Comment:
I'd love to see a conclusion containing a summary and a call to action with some links :-)